### ISSUE TITLE
Support CAA record type

### ISSUE THREAD

**Issue by siwyd** [terraform-providers/terraform-provider-ns1#10]
NS1 supports CAA records now via the v2 web interface and via the API (but the API docs still need to be updated).

This is the info on CAA records and the NS1 API as provided to me by NS1 support:

```
We have supported CAA record management through our API already and will update our API doc soon.

In the meanwhile, you can use our Portal or follow this example:

$ curl -X PUT -H 'X-NSONE-KEY: <YOUR_API_KEY>' 'https://api.nsone.net/v1/zones/example.test/caa.example.test/CAA' -d @JSON_FILE

with JSON_FILE's content:

{
  "zone": "example.test",
  "domain": "caa.example.test",
  "answers": [
    {
      "answer": [
        0,
        "issue",
        "ca.example.test"
      ]
    }
  ],
  "type": "CAA"
}

GET, POST and DELETE requests are similar to other record types.

You can find more information about CAA record at https://ns1.com/articles/caa-records.
```

I've tried simply adding the CAA record type to the list of supported record types in the provider and rebuilding the provider. This allows me to create the following record perfectly, but as far as I can tell there is an issue retrieving the record info from NS1 and parsing it.

This is the resource:
```
resource "ns1_record" "wydooghe_com_caa" {
  zone = "wydooghe.com"
  domain = "wydooghe.com"
  type = "CAA"

  answers = {
    answer = "0 issue letsencrypt.org"
  }

  answers = {
    answer = "0 issuewild ;"
  }
}
```

This is the Terraform output:
```
Error applying plan:

1 error(s) occurred:

* module.ns1.ns1_record.wydooghe_com_caa: 1 error(s) occurred:

* ns1_record.wydooghe_com_caa: json: cannot unmarshal number into Go struct field Answer.answer of type string

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

The response when calling the API via curl is the following:
```
❯ curl -X GET -H "X-NSONE-KEY: $NS1_APIKEY" 'https://api.nsone.net/v1/zones/siwyd.com/siwyd.com/CAA' | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   363    0   363    0     0    363      0 --:--:-- --:--:-- --:--:--   573
{
  "domain": "siwyd.com",
  "zone": "siwyd.com",
  "use_client_subnet": true,
  "answers": [
    {
      "answer": [
        0,
        "issuewild",
        ";"
      ],
      "meta": {},
      "id": "59a6e940c9c79d000120d3e3"
    },
    {
      "answer": [
        0,
        "issue",
        "letsencrypt.org"
      ],
      "meta": {},
      "id": "59a6e940c9c79d000120d3e4"
    }
  ],
  "id": "59a6e940c9c79d000120d3e5",
  "regions": {},
  "meta": {},
  "link": null,
  "filters": [],
  "ttl": 3600,
  "tier": 1,
  "type": "CAA",
  "networks": [
    0
  ]
}
```
The first element in the array is an integer, I'm guessing that might be the problem? MX records have a similar structure, but it seems NS1 returns those as all strings instead of the first element as integer:
```
❯ curl -X GET -H "X-NSONE-KEY: $NS1_APIKEY" 'https://api.nsone.net/v1/zones/siwyd.com/siwyd.com/MX' | jq . 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   362    0   362    0     0    362      0 --:--:-- --:--:-- --:--:--   500
{
  "domain": "siwyd.com",
  "zone": "siwyd.com",
  "use_client_subnet": true,
  "answers": [
    {
      "answer": [
        "10",
        "mxb.mailgun.org"
      ],
      "meta": {},
      "id": "59517182a632f6000189aff8"
    },
    {
      "answer": [
        "10",
        "mxa.mailgun.org"
      ],
      "meta": {},
      "id": "59517182a632f6000189aff9"
    }
  ],
  "id": "59517182a632f6000189affa",
  "regions": {},
  "meta": {},
  "link": null,
  "filters": [],
  "ttl": 3600,
  "tier": 1,
  "type": "MX",
  "networks": [
    0
  ]
}
```

My Golang knowledge (and Terraform provider debugging) is pretty much nonexistent, so any help to get this working is much appreciated ;)

**Comment by mburtless** [comment]
This is due to an issue with the NS1 SDK's inability to handle polymorphism in the API's responses (https://github.com/ns1/ns1-go/issues/38).  Will require a resolution in the SDK before this provider can support CAA records.

**Comment by ChrisMcKee** [comment]
With the ability to stuff json into a payload with tf12 is this any more fixable / workaroundable?
https://github.com/ns1/ns1-go/issues/38#issuecomment-516899945

Happily CAA records dont update that often but at the moment we run failover to R53; so we have 1 scripted and one where we have to remind people they need to manually login to the UI and fight the weird interface; and after 2 years it takes some remembering 😆

**Comment by mburtless** [comment]
Not that I know of, sadly.  Demarshalling the `int` in the JSON response via ns1-go is the core issue here.  If we want to avoid that and return the raw JSON to Terraform, we would still need to implement some custom code that returns `json.RawMessage` or something.  At that point, implementing proper polymorphism support is probably the better option IMO.

Addressing this is in our backlog (though may be delayed in favor of publishing a new version of the SDK which addresses this issue, among other things).  As always, PRs are welcome!

**Comment by rupa** [comment]
We've merged a change for this to the Go-SDK, working on updating the provider to support this.

**Comment by ChrisMcKee** [comment]
Ta chap

**Comment by ChrisMcKee** [comment]
https://github.com/terraform-providers/terraform-provider-ns1/pull/78

**Comment by rupa** [comment]
CAA support is merged and will be in `v1.5.4`
